### ISSUE TITLE
0.1.0 won't compile due to textnonce dependency

### ISSUE THREAD

**Issue by noam87** [iron/params#13]
compilation of `params` fails on `textnonce` dependency. Platform: Linux 64 bit (Ubuntu):

```Compiling textnonce v0.2.4
/home/computer/Dropbox/Code/dupernote/core/target/debug/build/textnonce-b46a6516de072497/out/lib.rs:97:26: 98:98 error: no method named`deserialize_newtype_struct`found for type`&mut __D`in the current scope
/home/computer/Dropbox/Code/dupernote/core/target/debug/build/textnonce-b46a6516de072497/out/lib.rs:97             deserializer.deserialize_newtype_struct("TextNonce",
/home/computer/Dropbox/Code/dupernote/core/target/debug/build/textnonce-b46a6516de072497/out/lib.rs:98                                                     __Visitor::<__D>(::std::marker::PhantomData))
/home/computer/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.4/src/lib.rs:16:1: 16:47 note: in this expansion of include!
/home/computer/Dropbox/Code/dupernote/core/target/debug/build/textnonce-b46a6516de072497/out/lib.rs:106:20: 106:66 error: no method named`serialize_newtype_struct`found for type`&mut __S`in the current scope
/home/computer/Dropbox/Code/dupernote/core/target/debug/build/textnonce-b46a6516de072497/out/lib.rs:106         serializer.serialize_newtype_struct("TextNonce", &self.0)
                                                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/computer/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.4/src/lib.rs:16:1: 16:47 note: in this expansion of include!
error: aborting due to 2 previous errors
Could not compile`textnonce`.

To learn more, run the command again with --verbose.

```
```

**Comment by RoxasShadow** [comment]
Same here, but with a different error:

``` rust
   Compiling formdata v0.7.10
   Compiling textnonce v0.2.6
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15:33: 15:46 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`) [E0308]
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15         serde_codegen::register(&mut registry);
                                                                                                                            ^~~~~~~~~~~~~
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15:33: 15:46 help: run `rustc --explain E0308` to see a detailed explanation
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17:33: 17:46 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`) [E0308]
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17         serde_codegen::register(&mut registry);
                                                                                                                            ^~~~~~~~~~~~~
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17:33: 17:46 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to previous error
Build failed, waiting for other jobs to finish...
error: aborting due to previous error
Could not compile `formdata`.
```

Probably related: https://github.com/mikedilger/formdata/issues/27

I also tried to upgrade the `formdata` version that `params` requires to the latest available one, but the problem still appears:

``` rust
   Compiling formdata v0.8.0
   Compiling textnonce v0.3.0
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.8.0/build.rs:15:33: 15:46 /Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.3.0/build.rs:17:33: 17:46 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`) [E0308]
error: /Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.8.0/build.rs:15         serde_codegen::register(&mut registry);
mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`)                                                                                                                            [E0308^~~~~~~~~~~~~]

/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.8.0/build.rs:15:33: 15:46 help: run `rustc --explain E0308` to see a detailed explanation/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.3.0/build.rs:
17         serde_codegen::register(&mut registry);
                                                                                                                            ^~~~~~~~~~~~~
/Users/giovanni/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.3.0/build.rs:17:33: 17:46 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to previous error
error: aborting due to previous error
Build failed, waiting for other jobs to finish...
Could not compile `textnonce`.

To learn more, run the command again with --verbose.
```

**Comment by SkylerLipthay** [comment]
So I believe this is most related to mikedilger/textnonce#7. Until that is sorted out, you can try your hand at `cargo update` or manually modifying `Cargo.lock` to resolve the situation hackishly.

**Comment by RoxasShadow** [comment]
`cargo update` doesn't produce any effect; if I modify manually `Cargo.lock`, it is reverted as soon as I run `cargo build`. However, it seems that cloning `formdata` too asking it to use `textnonce` `0.2.6` doesn't work either.

**Comment by Manicqin** [comment]
Same Problem as RoxasShadow, Tried the same things and failed... next stop is to dive into the code and understand the problem.

**Comment by ollivera** [comment]
Same Problem as RoxasShadow. I tried to play with the dependencies without success :(
I disabled params temporally otherwise I can't build my project.

**Comment by SkylerLipthay** [comment]
`params` is working in my current projects without issue. Can someone provide a full, minimal reproduction of this problem (a `Cargo.toml`)? I'll try my hand at this again. Apologies for the inconvenient delay!

**Comment by ollivera** [comment]
Hi, It is very easy to reproduce it. I am using rust 1.8.0. The steps are below:
1. Create a new project:

cargo new paramtest --bin
1. Add these dependencies:

Cargo.toml

[package]
name = "paramtest"
version = "0.1.0"
authors = ["ollivera"]

[dependencies]
hyper = "0.9.1"
iron = "0.3.0"
params = "0.1.0"
1. copy https://github.com/iron/params/blob/master/examples/params.rs to main.rs
2. Build: cargo build --release

/root/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15:33: 15:46 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`) [E0308]
/root/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15         serde_codegen::register(&mut registry);
                                                                                                                  ^~~~~~~~~~~~~
/root/.cargo/registry/src/github.com-88ac128001ac3a9a/formdata-0.7.10/build.rs:15:33: 15:46 help: run `rustc --explain E0308` to see a detailed explanation
/root/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17:33: 17:46 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut inner::syntex::Registry`
(expected struct `syntex::Registry`,
    found struct `inner::syntex::Registry`) [E0308]
/root/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17         serde_codegen::register(&mut registry);
                                                                                                                  ^~~~~~~~~~~~~
/root/.cargo/registry/src/github.com-88ac128001ac3a9a/textnonce-0.2.6/build.rs:17:33: 17:46 help: run `rustc --explain E0308` to see a detailed explanation

**Comment by SkylerLipthay** [comment]
Thanks!

Wow, I'm really feeling the pain of exposed/reexported types. If I update the `formdata` dependency to 0.8.1 in `params`, there exists an incompatibility issue between `iron`'s `hyper` 0.8 and `formdata`'s `hyper` 0.9. This would be fixed (temporary, probably) by iron/iron#452.

I'm sad to say that I'm finding the `formdata` dependency to be more and more problematic as time goes on. It might be best to switch to the less dependent crate, [cybergeek94/multipart](https://github.com/cybergeek94/multipart). This would resolve the everlasting `syntex` problems as well as these new `hyper` incompatibility problems, I believe.

The only item that `params` reexports from `formdata` is `formdata::Error`, so the switch to using `multipart` will be breaking but shouldn't be too hard to make. I'll look into this migration shortly to make sure it's a reasonable choice.

I'm very sorry to everyone for the inconvenience here! I've been busy and neglecting this project. I wish I had a temporary workaround to offer via editing `Cargo.lock`, but I can't seem to figure one out.

**Comment by SkylerLipthay** [comment]
I've released `0.2.0`! There were some small breaking changes that have been documented in this commit: 6362b6ebc4acbaea2eff5c1ae1c1a794576d45e7.

Thank you all for being patient. Please reopen this issue and let me know if there are any further problems. :heart:
